File: packages/core/tests/fixtures/typescript/code/classes/inheritance.ts
/**
 * Class inheritance and abstract classes
 * Tests: abstract class, extends, method overriding
 */

abstract class Animal {
  constructor(public name: string) {}

  abstract make_sound(): string;

  move(distance: number = 0): void {
    console.log(`${this.name} moved ${distance} meters`);
  }
}

class Dog extends Animal {
  constructor(name: string) {
    super(name);
  }

  make_sound(): string {
    return "Woof!";
  }

  fetch(item: string): void {
    console.log(`${this.name} fetched ${item}`);
  }
}

class Cat extends Animal {
  make_sound(): string {
    return "Meow!";
  }
}

export { Animal, Dog, Cat };

File: packages/core/tests/fixtures/python/code/classes/inheritance.py
"""
Class inheritance
Tests: inheritance, super(), method overriding
"""

class Animal:
    def __init__(self, name: str):
        self.name = name

    def speak(self) -> str:
        raise NotImplementedError("Subclass must implement speak()")

    def move(self, distance: int = 0):
        print(f"{self.name} moved {distance} meters")


class Dog(Animal):
    def __init__(self, name: str, breed: str):
        super().__init__(name)
        self.breed = breed

    def speak(self) -> str:
        return "Woof!"

    def fetch(self, item: str):
        print(f"{self.name} fetched {item}")


class Cat(Animal):
    def speak(self) -> str:
        return "Meow!"

    def climb(self):
        print(f"{self.name} is climbing")
